### ISSUE TITLE
Output batching with sparse child array mapping

### ISSUE THREAD

**Issue by Archanian** [benthosdev/benthos#991]
Hello! I'm seeking some guidance on output batching for my specific use case, having a little difficulty figuring out how to get output batching working as I would like (config example below). Using a `kafka` input, essentially each input message has a `Children` array that can contain 0-N elements, and this is generally quite sparse in distribution. On the high end this array contains ~200 elements, for many messages it contains only a few, and for at least 50% it is an empty collection.

The topic being consumed has 30 partitions, and this pipeline has 5 brokered input copies running across 6 replicas (so each input copy has a single partition).

```
input:
  broker:
    copies: 5
    inputs:
      - kafka:
          addresses:
            - ${KAFKA_BROKERS}
          consumer_group: ${CONSUMER_GROUP}
          topics:
            - ${KAFKA_TOPIC}
          batching:
            period: 100ms
            count: 5

pipeline:
  processors:
    - bloblang: |
        root = this.Children.map_each(
          x -> {
            "ID": x.Id,
            "Link": x.Url
          })
    - unarchive:
        format: json_array

output:
  broker:
    copies: 10
    pattern: greedy
    outputs:
      - aws_dynamodb:
          backoff:
            initial_interval: 1s
            max_interval: 30s
          max_retries: 5
          max_in_flight: 3
          table: "<table_name>"
          json_map_columns:
            "": .
```

I can get this working if I define batching on the input (as seen above), but then this results in very spiky output batches (ie. up to ~200 items in batch writes) which slows down throughput _massively_ (I assume the dynamo output hits the 25 max batch write limit, and writes the items one-by-one).

If I define batching on the output, processing stalls completely, since I assume the input is waiting to commit the offset for a batch that will never be flushed in the case of the processor not generating any messages (ie. all messages in the input batch have an empty Children array).

What I really want is to be able to limit the output batch to 25 items (falls within Dynamo batch write limit) somehow accounting for the fact that input items are not guaranteed to produce any messages. Is there a way to achieve this? I assume I'm missing something critical in the way I define the input, but I'm kinda stuck right now.

Any guidance is much appreciated!

**Comment by Archanian** [comment]
I think I may have solved at least part of this with the `split` processor ... is there anything else I can do at the input level to improve throughput?

**Comment by Jeffail** [comment]
Hey @Archanian, there's a few things that might help.

If you're using a relatively up to date version of benthos you don't need brokers on either end, the input will automatically saturate processing threads and outputs, and I believe the output should perform just as well by removing the broker and increasing the field `max_in_flight`. 

The next thing that'll help is that you can make the batching simpler now by setting the kafka field [`checkpoint_limit`](https://www.benthos.dev/docs/components/inputs/kafka#checkpoint_limit). This field controls how many messages of the same partition can be processed asynchronously (out of order), and allows messages to flow through to the output level without the awkward lockstep that forces you to batch at the input level.

If you set that to something liberal like 100 then you'll be able to batch messages at the output level and even out the flow, as the batching will be satisfied both by slow traffic of messages with lots of items and also high traffic of messages with fewer items.

Finally, this is a preference thing but I would perform the unarchive bit at the input level and end it with a split so that the batches are fully broken down before the processing pipelines. You don't have any other processing right now but if you were to add some in the future then this would help vertically scale the messages across processing threads.

With all that stuff together I think this config might be a good place to start:

```yaml
input:
  kafka:
    addresses:
      - ${KAFKA_BROKERS}
    consumer_group: ${CONSUMER_GROUP}
    topics:
      - ${KAFKA_TOPIC}
    checkpoint_limit: 100
  processors:
    - bloblang: |
        root = this.Children.map_each(
          x -> {
            "ID": x.Id,
            "Link": x.Url
          })
    - unarchive:
        format: json_array
    - split: {}

output:
  aws_dynamodb:
    backoff:
      initial_interval: 1s
      max_interval: 30s
      max_retries: 5
    max_in_flight: 30
    table: "<table_name>"
    json_map_columns:
      "": .
    batching:
      period: 500ms
      count: 20
```

Let me know if that seems to work better. In the next major version release I'm going to be re-evaluating the default values we have for things like `max_in_flight` and `checkpoint_limit`, so I'm quite interested in feedback.

**Comment by Archanian** [comment]
Hey @Jeffail thanks for the prompt and detailed response, I really appreciate it. One question I have about increasing `checkpoint_limit` to 100 - would this still result in stalling if there were no children elements in a 100 message span? Would the limit need to be such that there is a guarantee of at least one message? If so, that presents a problem, since the theoretical limit is infinite.

Good to know about the brokers, I'll test that out and see what the throughput difference is. We're also running these pipes in ECS containers, so the CPU scheduling is a little less reliable. You raise a good point about scaling processors, I'll move the existing ones to the input.

With the original config I posted, plus a `split` processor to ensure batches are limited to 25, I have been able to achieve ~60k writes/sec to Dynamo, which I think is pretty good? It could probably go even faster with bigger containers, CPU was close to saturated.

**Comment by Jeffail** [comment]
No problem @Archanian! You won't get any issues with stalling. One thing that might be worth also doing if the number of children might be zero is delete the messages from the bloblang mapping with an assert like this:

```yaml
        root = this.Children.map_each(
          x -> {
            "ID": x.Id,
            "Link": x.Url
          }).not_empty().catch(deleted())
```

**Comment by Archanian** [comment]
OK so something is not right here ... I have made the changes you recommended - removed brokers, added `checkpoint_limit` of 100 to the input, and adding batching to the output, ie:
```
output:
  aws_dynamodb:
    batching:
      period: 200ms
      count: 25
    backoff:
      initial_interval: 1s
      max_interval: 30s
    max_retries: 5
    max_in_flight: 30
    table: "<table_name>"
    json_map_columns:
      "": .
```

This prevents stalling, however it appears batching is not working correctly, I'm seeing batches wayyyy over 25 (and as a result, terrible throughput). Adding a `split` processor with a size of 25 has the effect of _partially_ limiting batches, but not all them, its really quite strange ...

![image](https://user-images.githubusercontent.com/5627057/144927187-bcb7edd4-bb57-4fcf-a0b9-b9ca9ae0d669.png)

As you can see, a lot of batches are capped at 25, but a bunch of them aren't. It doesn't make any sense.

As a side note, it would be great if the dynamo writer would internally cap batch write operations at 25 items.

**Comment by Archanian** [comment]
@Jeffail as another side note, I didn't mention this earlier but we are specifying a thread count (20) on the pipeline which results in much greater throughput. Moving the processor to the input means we don't have this control, and throughput is much lower a result.

**Comment by Jeffail** [comment]
@Archanian yeah that's odd, can you share the whole config? I'll try and reproduce it locally.

**Comment by Archanian** [comment]
Sure thing:

```
input:
  kafka:
    addresses:
      - ${KAFKA_BROKERS}
    checkpoint_limit: 100
    consumer_group: ${CONSUMER_GROUP}
    topics:
      - ${KAFKA_TOPIC}

pipeline:
  threads: 20
  processors:
    - bloblang: |
        root = this.Children.map_each(
          l -> {
            "ID": l.Id
            "Link": l.Url,
          }).flatten()
    - unarchive:
        format: json_array
    - split:
        size: 25 

output:
  aws_dynamodb:
    batching:
      period: 200ms
      count: 25
    backoff:
      initial_interval: 1s
      max_interval: 30s
    max_retries: 5
    max_in_flight: 30
    region: us-east-2
    table: "<table_name>"
    json_map_columns:
      "": .
```

Remove the `split` processor for _truly_ wild batch sizes :D

**Comment by Jeffail** [comment]
Ah okay, does everything look good when you run with:

```yaml
input:
  kafka:
    addresses:
      - ${KAFKA_BROKERS}
    checkpoint_limit: 100
    consumer_group: ${CONSUMER_GROUP}
    topics:
      - ${KAFKA_TOPIC}

pipeline:
  threads: 20
  processors:
    - bloblang: |
        root = this.Children.map_each(
          l -> {
            "ID": l.Id
            "Link": l.Url,
          }).flatten()
    - unarchive:
        format: json_array
    - split:
        size: 1

output:
  aws_dynamodb:
    batching:
      period: 200ms
      count: 25
    backoff:
      initial_interval: 1s
      max_interval: 30s
    max_retries: 5
    max_in_flight: 30
    region: us-east-2
    table: "<table_name>"
    json_map_columns:
      "": .
```

**Comment by Archanian** [comment]
That's done the trick, batch sizes look correct now. So the split of 1 is required in the processor pipeline?

**Comment by Jeffail** [comment]
Yeah so the reason for that is that batching mechanisms are inclusive of the batch that triggered it, so if you have the config:

```yaml
    batching:
      count: 25
```

And then feed it a batch of 2 messages, then followed by a batch of 30 messages, you'll end up with a flushed batch of 32 messages. This behaviour is intentional as we consider batches to be sacred and so we keep them together unless your configuration instructs otherwise, that's why you need to split the batch resulting from `unarchive` explicitly.

There's definitely room for reconsidering this as the default, where we could add an opt-in configuration to batching mechanisms that preserves the input batches, but we'd need to wait for v4 before we can implement it as it'd be a breaking change.

However, I can definitely add some debug logs to the batching mechanism so that it's clearer how and why these batches are being formed the way that they are. I'm going to mark this as a documentation issue to get that bit done, but if you have further questions then feel free to continue the convo in here.

**Comment by Archanian** [comment]
Gotcha, thanks for the explanation!

What are your thoughts on making the dynamo writer internally limit bulk writes to 25 items? So it would split any batches larger than this making it invisible to the end user.

This has a huge impact on throughput from my tests, it might be worth considering?

**Comment by Jeffail** [comment]
Yeah good point, definitely something we should be forcing at the lower level, can you open a separate issue for that?

**Comment by Archanian** [comment]
Absolutely, [linked here](https://github.com/Jeffail/benthos/issues/992).
